### ISSUE TITLE
Some questions about SHOT code.

### ISSUE THREAD

**Issue by EricKani** [tim-learn/SHOT#18]
Hi Tim,

Firstly thanks for sharing the code to the community. I encountered some questions in my code reading process.

In your "/object/image_target.py" file:
1) **Line 191**: the "_tar_idx_" here is a batch of dataloader of "_dset_loaders["target"]_", but "_mem_label_" is the prediction of "_dset_loaders['test']_". Are they matching in content~?
2) **Line 192**: is there a softmax operator needed for "outputs_test"? I found there is a \dleta in equ.7.
3) Line 264 and 265 are not been used below.
4) **Line 270**: why should we concatenate a "torch.ones(all_fea.size(0), 1))" to our "all_fea" in channel dimension when "distance" is "cosine"~? I don't understand this...

Thanks!

Eric

**Comment by tim-learn** [comment]
> Hi Tim,
> 
> Firstly thanks for sharing the code to the community. I encountered some questions in my code reading process.
> 
> In your "/object/image_target.py" file:
> 
> 1. **Line 191**: the "_tar_idx_" here is a batch of dataloader of "_dset_loaders["target"]_", but "_mem_label_" is the prediction of "_dset_loaders['test']_". Are they matching in content~?
> 2. **Line 192**: is there a softmax operator needed for "outputs_test"? I found there is a \dleta in equ.7.
> 3. Line 264 and 265 are not been used below.
> 4. **Line 270**: why should we concatenate a "torch.ones(all_fea.size(0), 1))" to our "all_fea" in channel dimension when "distance" is "cosine"~? I don't understand this...
> 
> Thanks!
> 
> Eric

Thanks, Eric!

1. the test loader and the target loader are consistent, differs only in the data transformation function.
2.  for crossentropyloss, the softmax is hidden inside.
3.  yes, lines 264-265 are not used, you can ignore them.
4. you can refer #13 for this question.

Best, 

Tim

**Comment by EricKani** [comment]
Thanks for your quick response, Tim!

I have understand the first three answers.
I will think about the fourth question further based on # 13 issue.

Best,
Eric
